calculator_test.go

```diff
@@ -2,23 +2,189 @@ package calculator_test
+	var testCases = []struct {
+		name string
+		nums []float64
+		want float64
+	}{
+		{"positive", []float64{2, 2}, 4},
+		{"negative", []float64{-2, 5}, 3},
+		{"one zero", []float64{0, 3}, 3},
+		{"two zeroes", []float64{0, 0}, 0},
+		{"decimal", []float64{2.5, 4}, 6.5},
+		{"empty slice", []float64{}, 0},
+		{"one item in slice", []float64{3}, 3},
+		{"many numbers", []float64{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 55},
+	}
+	for _, tc := range testCases {
+		got := calculator.Add(tc.nums...)
+		if tc.want != got {
+			t.Errorf("test name: %s, want %f, got %f", tc.name, tc.want, got)
+		}
+	}
+}
+
+func TestAddRandom(t *testing.T) {
+	t.Parallel()
+	rand.Seed(time.Now().UnixNano())
+
+	for i := 0; i < 1000; i++ {
+		a, b := rand.Float64(), rand.Float64()
+		want := a + b
+		got := calculator.Add(a, b)
+		if want != got {
+			t.Errorf("adding %f, %f: want %f, got %f", a, b, want, got)
+		}
+	}
+}
+
+func TestSubtract(t *testing.T) {
+	t.Parallel()
+	var testCases = []struct {
+		name string
+		nums []float64
+		want float64
+	}{
+		{"positive", []float64{2, 2}, 0},
+		{"negative", []float64{-2, 5}, -7},
+		{"one zero", []float64{0, 3}, -3},
+		{"two zeroes", []float64{0, 0}, 0},
+		{"decimal", []float64{2.5, 4}, -1.5},
+		{"empty slice", []float64{}, 0},
+		{"one item in slice", []float64{3}, 3},
+		{"many numbers", []float64{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, -53},
+	}
+	for _, tc := range testCases {
+		got := calculator.Subtract(tc.nums...)
+		if tc.want != got {
+			t.Errorf("test name: %s, want %f, got %f", tc.name, tc.want, got)
+		}
+	}
+}
+
+func TestMultiply(t *testing.T) {
+	t.Parallel()
+	var testCases = []struct {
+		name string
+		nums []float64
+		want float64
+	}{
+		{"positive", []float64{2, 2}, 4},
+		{"negative", []float64{-2, 5}, -10},
+		{"one zero", []float64{0, 3}, 0},
+		{"two zeroes", []float64{0, 0}, 0},
+		{"decimal", []float64{2.5, 4}, 10},
+		{"empty slice", []float64{}, 0},
+		{"one item in slice", []float64{3}, 3},
+		{"many numbers", []float64{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 3628800},
+		{"many of the same numbers", []float64{2, 2, 2, 2, 2, 2, 2, 2, 2, 2}, 1024},
+	}
+	for _, tc := range testCases {
+		got := calculator.Multiply(tc.nums...)
+		if tc.want != got {
+			t.Errorf("test name: %s, want %f, got %f", tc.name, tc.want, got)
+		}
+	}
+}
+
+func TestDivide(t *testing.T) {
+	t.Parallel()
+	var testCases = []struct {
+		name        string
+		nums        []float64
+		want        float64
+		errExpected bool
+	}{
+		{"positive", []float64{2, 2}, 1, false},
+		{"negative", []float64{-2, 5}, -0.4, false},
+		{"one zero", []float64{0, 3}, 0, true},
+		{"two zeroes", []float64{0, 0}, 0, true},
+		{"decimal", []float64{2.5, 4}, 0.625, false},
+		{"empty slice", []float64{}, 0, false},
+		{"one item in slice", []float64{3}, 3, false},
+		{"many numbers", []float64{1024, 2, 2, 2, 2, 2, 2, 2, 2, 2}, 2, false},
+	}
+	for _, tc := range testCases {
+		got, err := calculator.Divide(tc.nums...)
+		if tc.errExpected {
```

Comment:
And we can refactor this code to be more concise and handle _both_ unexpected error status cases, can't we? Can you see how? (Hint: suppose we had some `bool` variable `errReceived`, telling us whether or not we received an error from `Divide`. That would make the `if` statement easy to write, wouldn't it? How could we set such a variable?)

Reply:
Thanks for the feedback.

I'm not sure I completely understand what you're asking here.

Do we want to assert specific errors here, coming from upstream functions (in this case, `Divide()`)? 

In that case, one way that comes to mind is to make `StringMath()` return the extra `bool` value that tells us whether errors are from itself or from other functions it is calling.

It feels like extra machinery though...

Reply:
I tried to implement this in [this commit](https://github.com/micuffaro/ftl-fundamentals/pull/1/commits/e41009eb2149320aaa62c3a7b21dd902d968c63b), although frankly I don't see how my approach is adding anything useful to the current `StringMath()` test, as I can probably leave it at asserting the `!tc.errExpected` block.